### PR TITLE
Add undo functionality

### PR DESCRIPTION
**Pull request by kckrithika**

### CODE REVIEW

File: GameHost/app.py [hunk 1]
```diff
@@ -5,15 +5,20 @@
...
+import os
 from flask_cors import CORS, cross_origin
 import jsonpickle
 
 app = Flask(__name__)  # create an app instance
 CORS(app, resources={r"/*": {"origins": "*"}})
 app.config['CORS_HEADERS'] = 'Content-Type'
+app.config['SESSION_PERMANENT'] = True
+app.config['SESSION_TYPE'] = 'filesystem'
+app.config['PERMANENT_SESSION_LIFETIME'] = timedelta(minutes=30)
```
Comment (rama56) [inline GameHost/app.py]: Why 30 minutes? What does this do?

File: GameHost/app.py [hunk 2]
```diff
@@ -5,15 +5,20 @@
...
+app.secret_key = os.urandom(16)
```
Comment (rama56) [inline GameHost/app.py]: what is this 16?

File: GameHost/app.py [hunk 3]
```diff
@@ -26,38 +31,55 @@ def create_new_game_test():
...
 
 @app.route("/newgame")  # at the end point /newgame
+@cross_origin(supports_credentials=True)
 def create_new_game():
 
     game_state = GameState()
-
     # Use JSON pickle to convert to JSON. Native json module can't handle complex classes.
     json_response = jsonpickle.encode(game_state)
+    session['history'] = [jsonpickle.encode(game_state)]
```
Comment (rama56) [inline GameHost/app.py]: Use json_respone from above line.

File: GameHost/app.py [hunk 4]
```diff
@@ -26,38 +31,55 @@ def create_new_game_test():
...
     game_state = GameState()
     game_state.populate_test_values()
-
     # Use JSON pickle to convert to JSON. Native json module can't handle complex classes.
     json_response = jsonpickle.encode(game_state)
     return json_response
 
 
 @app.route("/newgame")  # at the end point /newgame
+@cross_origin(supports_credentials=True)
```
Comment (rama56) [inline GameHost/app.py]: What is this for? I thought CORS issue was resolved last time.

File: GameHost/app.py [hunk 5]
```diff
@@ -26,38 +31,55 @@ def create_new_game_test():
...
 @app.route("/restart")  # at the end point /restart
+@cross_origin(supports_credentials=True)
 def restart():
     game_state = GameState()
-
+    session['history'] = []
     game_state.restart()
     # Use JSON pickle to convert to JSON. Native json module can't handle complex classes.
 
     return game_state
```
Comment (rama56) [inline GameHost/app.py]: Convert to JSON before sending to the client.

File: GameHost/app.py [hunk 6]
```diff
@@ -26,38 +31,55 @@ def create_new_game_test():
...
     # Use JSON pickle to convert to JSON. Native json module can't handle complex classes.
     json_response = jsonpickle.encode(game_state)
+    session['history'] = [jsonpickle.encode(game_state)]
     return json_response
 
 
 @app.route("/restart")  # at the end point /restart
+@cross_origin(supports_credentials=True)
 def restart():
     game_state = GameState()
```
Comment (rama56) [inline GameHost/app.py]: This is wrong. This is 'new game' functionality.

File: GameHost/app.py [hunk 7]
```diff
@@ -26,38 +31,55 @@ def create_new_game_test():
...
     # Use JSON pickle to convert to JSON. Native json module can't handle complex classes.
 
     return game_state
 
 
+@app.route("/undo")
+@cross_origin(supports_credentials=True)
+def undo():
+    if ('history' not in session.keys()) or (len(session['history']) == 0):
+        return ""
```
Comment (rama56) [inline GameHost/app.py]: What does the client do with this empty string?

File: GameHost/app.py [hunk 8]
```diff
@@ -26,38 +31,55 @@ def create_new_game_test():
...
-    state_dict = request.json
     # Use JSON pickle to convert JSON to Python object. Native json module can't handle complex classes.
-    game_state = jsonpickle.decode(json.dumps(state_dict))
+    game_state = jsonpickle.decode(json.dumps(request.json))
     game_state.alter_state()
+    session['history'].append(jsonpickle.encode(game_state))
+    session.modified = True
 
     # Use JSON pickle to convert to JSON. Native json module can't handle complex classes.
     json_response = jsonpickle.encode(game_state)
```
Comment (rama56) [inline GameHost/app.py:124]: Move this line to after  "game_state.alter_state()" and use it in the subsequent line to append to history.

File: GameHost/app.py [hunk 9]
```diff
@@ -26,38 +31,55 @@ def create_new_game_test():
...
+    session['history'] = [jsonpickle.encode(game_state)]
     return json_response
 
 
 @app.route("/restart")  # at the end point /restart
+@cross_origin(supports_credentials=True)
 def restart():
     game_state = GameState()
-
+    session['history'] = []
```
Comment (rama56) [inline GameHost/app.py]: Why assign it an empty list?

File: GameHost/app.py [hunk 10]
```diff
@@ -26,38 +31,55 @@ def create_new_game_test():
...
     game_state.restart()
     # Use JSON pickle to convert to JSON. Native json module can't handle complex classes.
 
     return game_state
 
 
+@app.route("/undo")
+@cross_origin(supports_credentials=True)
+def undo():
```
Comment (rama56) [inline GameHost/app.py:68]: If possible, add an Unit Test.

**Review by rama56 — CHANGES_REQUESTED** [review]

File: GameHost/app.py (hunk 1 above)
Comment (kckrithika) [reply GameHost/app.py]: It retains the session for 30 minutes

File: GameHost/app.py (hunk 2 above)
Comment (kckrithika) [reply GameHost/app.py]: 16 digit secret key, made of random characters

File: GameHost/app.py (hunk 4 above)
Comment (kckrithika) [reply GameHost/app.py]: CORS issue was solved by allowing access to specified resources from specified origins. To enable sessions, flask requires the client to send credentials. That's why the 'xmlhttp.withCredentials' was set in the front end code. To receive the credentials, we need to allow it explicitly.

File: GameHost/app.py (hunk 7 above)
Comment (kckrithika) [reply GameHost/app.py]: client is unable to parse this. It's not handled gracefully. Will make the UI handle it gracefully in the next cut. For now, it does not affect any of the existing functionalities

File: GameHost/app.py (hunk 10 above)
Comment (rama56) [reply GameHost/app.py:68]: Also add into README.md the issues faced and how it were solved.

**Pull request merged**